Feature: Add Alpha and Beta metrics to backtesting summary

ccf opened the pull request:
## Summary
Adds Alpha and Beta metrics to backtesting summary, providing market-relative performance analysis.

## Motivation
Provides users with metrics to understand if their strategy adds value beyond passive market exposure. Alpha measures excess returns and Beta measures market sensitivity.

## Sources

The canonical sources from which the Alpha and Beta derivations originate:

1. William Sharpe (1964) - "Capital Asset Prices: A Theory of Market Equilibrium under Conditions of Risk", The Journal of Finance, Vol. 19, No. 3, pp. 425-442
  * Nobel Prize-winning work that introduced CAPM
2. John Lintner (1965) - "The Valuation of Risk Assets and the Selection of Risky Investments in Stock Portfolios and Capital Budgets", The Review of Economics and Statistics, Vol. 47, No. 1, pp. 13-37
  * Independent derivation of CAPM
3. Jan Mossin (1966) - "Equilibrium in a Capital Asset Market", Econometrica, Vol. 34, No. 4, pp. 768-783
  * Further theoretical development of CAPM
4. Michael C. Jensen (1968) in "The Performance of Mutual Funds in the Period 1945-1964", The Journal of Finance, Vol. 23, No. 2, pp. 389-416.
  * Alpha is simply the CAPM formula rearranged to solve for the excess return (alpha)

Investopedia articles providing summaries:

- https://www.investopedia.com/articles/financial-theory/08/deeper-look-at-alpha.asp
- https://www.investopedia.com/terms/b/beta.asp

## Changes

- Add `calculate_alpha_beta()` function using CAPM model
- Add helper functions for daily returns calculation
- Include `processed_data` in backtest results for market benchmark
- Display Alpha and Beta in SUMMARY METRICS table
- Fix typo: `sharp_ratio` → `sharpe_ratio`

## Features
- **Alpha**: Annualized excess return over market benchmark (buy-and-hold pairlist)
- **Beta**: Strategy sensitivity to market movements (volatility measure)
- Uses risk-free rate of 0% (consistent with existing Sharpe/Sortino)

## Testing
- 3 new test functions with parametrized test cases
- Tests cover edge cases (empty data, no market data, single trade)
- Tests validate Beta correlation patterns
- All tests follow existing patterns in test_btanalysis.py

## Backward Compatibility
- Fully backward compatible
- Graceful fallback if processed_data unavailable
- Defensive checks for missing Alpha/Beta in old results

AI Disclosure: AI was partially used to create this PR.

xmatthias commented on freqtrade/optimize/hyperopt_loss/hyperopt_loss_sortino.py:
```diff
@@ -34,6 +34,12 @@ def hyperopt_loss_function(
         """
-        sortino_ratio = calculate_sortino(results, min_date, max_date, starting_balance)
+        # Extract risk_free_rate from config if available
+        config = kwargs.get("config", {})
```
don't use kwargs - it's there to allow for breaking changes by adding new parameters - NOT to extract parameters that are used.

all arguments available to hyperopt_loss functions are available all the time (otherwise, they'll never be available) - so extracting from kwargs is not something we'll start doing.

ccf replied:
This is a good point. The interface explicitly defines config as a named parameter. Some of the hyperopt loss implementations use kwargs to accept parameters they don't use, but not to extract parameters they do use. I'll clean up.

xmatthias replied:
the interface defines an explicit parameter - but with an appropriate type alias.
`dict` (what you've used now) is NOT the right type.

Also - Please don't forget the main review comment - which does require answers before i look at this further.

xmatthias commented on freqtrade/data/metrics.py:
```diff
@@ -318,58 +318,99 @@ def calculate_expectancy(trades: pd.DataFrame) -> tuple[float, float]:
 
     if up_stdev != 0:
-        sharp_ratio = expected_returns_mean / up_stdev * np.sqrt(365)
+        # Annualize: multiply by sqrt(365)
```
your AI made pointless comments? 

I'd appreciate if you can do a review yourself and remove pointless comments like this (it's one - but very likely not the only one).

xmatthias commented on docs/backtesting.md:
````diff
@@ -417,6 +419,53 @@ It contains key metrics about the performance of your strategy on backtesting da
+}
+```
+
+This is optional and rarely needed for crypto backtesting, but can be useful for comparison with traditional financial markets or to model opportunity cost (e.g., stablecoin lending yields).
````
if it's rarely needed -  it's pointless to add to a crypto bot? (hence - it adds additional lines for no good reason?)

why add it in the first place, then? Take your pick - it's EITHER needed - or it isn't (and consequently should be removed from this PR).

ccf replied:
If your point is stylistic, I'm happy to adjust or remove the context altogether, if that's your preference. I'm saying that the average user won't care to adjust this unless they are trying to compare their strategy to strategies across other asset classes, or in more quantitatively rigorous settings.

Alpha, Sharpe, Sortino, and several other measures developed throughout modern portfolio theory consider what return one can earn without risk when benchmarking the returns you are getting while taking on risk. During ZIRP (2008-2015, 2020-2022), most of the finance world considered the risk-free return 0 because the yield on treasury bills was ~0% during that time. The current implementations of Sharpe and Sortino have the risk-free rate hardcoded as 0. This has not been the case for a while now -- 3 month treasury bills currently yield ~4.25%. If I want to accurately measure and compare the Alpha, Sharpe, and Sortino of my freqtrade strategy against other managers and asset classes, I need to change the risk-free rate of return. In fact, if I were to publish these these types of statistics formally about a strategy, I would need to change risk-free rate of return.

xmatthias replied:
stylistic ... not so much.

I take issue with 2 things in this PR (leading me to the edge of closing it, asking you to split it up).

1 - it's clearly an AI generated PR (at least in part) - which we ask people [to disclose](https://github.com/freqtrade/freqtrade/blob/8e44bc48b868d23a5ce6a938e736a920b9b3275c/CONTRIBUTING.md#L15) which you didn't
2 - it merges 2 features together, making a review of this pretty difficult (if i were to reject one part - do i close the PR outright even if the other part would be good?)

---

this is about the 2nd part, really - which you claim it's "rarely needed" - making me question if it's worth the about 200 lines you had to change (which are lines we have to maintain going forward) for it.
It blows the PR up beyond being reviewable in a reasonable time - because it merges together 2 different features.

I'm not against this _per se_ - but it should be a separate PR, not snuck into another PR trying to add new metrics.

it's 2 different topics - and while i understand that one may depend on the other - it complicates the review process a lot.

ccf replied:
I'll address all of your feedback and ensure the PR meets your guidelines as I work through each bit. Ultimately, I'm trying to contribute to and help the overall project. My initial implementation only introduced Alpha with a parameterized risk-free rate, but it felt short-sighted to not fix Sharpe and Sortino to be consistent. Assuming a fixed 0.0% risk-free rate is incorrect. William Sharpe won a Nobel Prize in Economics for his work on the Sharpe ratio and CAPM (what Alpha is based on), and he would not have included a risk-free rate variable if it was meant to just always be 0.0. If I should have done them in 2 separate PRs, my apologies -- it felt wrong to leave some metrics hardcoded while introducing a new one that properly parameterized it.

On the "optional" trigger: one can just use default (0.0) if they are testing changes to a given freqtrade strategy, but if they try to compare your own Sharpe, Sortino, or Alpha to that of other investments offered by institutional asset managers, their metrics will discount a real risk-free rate and thus be lower values. It won't be an apples to apples comparison from a modern portfolio construction perspective. A configurable risk-free rate allows users to adjust for this, if they need to in their case. This is why I called it "optional", but perhaps not the best term. It's really just a more advanced requirement that is relevant when trying to compare a freqtrade strategy to, say, a hedge fund or some other alternative investment.

xmatthias commented on docs/backtesting.md:
```diff
@@ -417,6 +419,53 @@ It contains key metrics about the performance of your strategy on backtesting da
+This is optional and rarely needed for crypto backtesting, but can be useful for comparison with traditional financial markets or to model opportunity cost (e.g., stablecoin lending yields).
+
+!!! Note
+    Alpha and Beta require historical market data for all pairs in your pairlist. They will show as "N/A" if this data is unavailable. The calculations use the Capital Asset Pricing Model (CAPM) formula.
```
Thats a pointless comment i think - and/or badly worded.

If you backtest - you MUST have historical data available for the pairs you backtest (otherwise they'll not be part of the backtest, anyway).

If this refers to "all data must be available for the whole time range" - then it's not realistic - as pairs are listed / unlisted all the time - so deviating timeranges between pairs are almost a given unless you only backtest for 3 pairs.
If this latter point is the case - then i don't see any check for this.

on top of that - the last sentence doesn't really relate to the first 2 - making this sound  pretty odd.

xmatthias reviewed (changes requested):
Unfortunately - as so many PR's recently, it feels like a rushed AI attempt without proper review ...  as it has contradicting statements in the docs - which essentially make half of the PR pointless (depending on how you impterpret that part).

There's no problem with usage of AI per se (it's pretty clear it is in some areas) - as long as usage is disclosed (which it wasn't), and you did a proper review yourself - which i feel didn't happen in some areas.

---

Also in reality, this is 2 features (add risk-free rate configurability, and new metrics).

Reviewing would be a LOT faster and easier if this were split into 2 different PR's - as now it mixes everything from formatting change and pointless comments (most likely not well reviewed AI changes) to actually new functionality (without links to verify the functionality is actually correct).

---

You're adding new metrics - but without source.
So  we'll have to trust that your calculation is actually correct? (tests will confirm that your implementation corresponds to tests - but not to the actual, commonly used understanding).

I'm sure i can find links for that - which probably deviate from what you implemented so it's best to include the links you used as source - at least in the description - to avoid discussion and doubts caused by different random google results used for the calculation definition.

---

I'll mark this as "request for changes" for a couple of  reasons (after a quick glance - not a proper, thorow review)

one - there's no links to the calculations you used - so we have no way to verify that your calculation matches common understanding (without googling for proably different definitions, anyway).

two - you're contradicting yourself - if risk-free rate is irrelevant for crypto trading - we don't need it as a parameter.
If it is - don't say it's irrelevant in the docs (and include a link or at least explanation to why it's relevant, maybe?).

there's also quite some cleanup to do (it feels like every line got a comment describing what it does, which is pretty pointless - and a pretty clear sign of AI at that).

xmatthias commented on freqtrade/optimize/hyperopt_loss/hyperopt_loss_sortino.py:
```diff
@@ -26,6 +26,7 @@ def hyperopt_loss_function(
         min_date: datetime,
         max_date: datetime,
         starting_balance: float,
+        config: dict,
```
Wrong type.

xmatthias commented on freqtrade/optimize/backtesting.py:
```diff
@@ -1735,6 +1735,7 @@ def backtest(
             "canceled_entry_orders": self.canceled_entry_orders,
             "replaced_entry_orders": self.replaced_entry_orders,
             "final_balance": self.wallets.get_total(self.strategy.config["stake_currency"]),
+            "processed_data": processed,
```
I'm not sure that's a good idea.
This is potentially large data - with the potential of causing a memory leak (especially relevant for hyperopt and backtesting in webserver mode).

It's also not actually necessary - optimize reports gets `btdata` already - which contains the raw data (it's not the same as processed - but you don't need the strategy's indicators).

xmatthias commented on freqtrade/data/metrics.py:
```diff
@@ -436,3 +461,113 @@ def calculate_sqn(trades: pd.DataFrame, starting_balance: float) -> float:
+    return alpha, beta
+
+
+def _calculate_daily_returns(
```
This whole method has an implicit problem (one you can't work around currently).
It says daily profit - but will ignore all open trades - and include "daily profit" the day it's realized.

On a trade with 3 month duration - you'll hence have 0 for 3 months, and then 10% (or whatever the profit was) on the day it closes.
Market will have moved in the meantime.

So combining this with daily data for trades will almost certainly result in a completely wrong result (for every strategy with holding times > 1 day) - as you combine daily profit (a value once every 3 months, in the above case) with the daily movement (essentially, with the movement for that particular day).

Your "combined" df will then look like this:
| date | strategy | market change|
| ----- | ------ | ------ |
| 2025-01-01 | 0 | +10 |
| 2025-01-02 | 0 | +5 |
| 2025-01-03 | 0 | -1 |
...
| 2025-03-01 | +10 | -1 |


Now i assume (didn't do the math for this sample) that this strategy will look great - as it had a 10% profit on a -1 market day ... but it ignores that the cumulative market may have gained 200% over this timeperiod.

There's a solution for that on the horizon (a feature i didn't complete so far) - which will provide accurate wallet data at a candle level during backtesting - which should allow this calculation to actually work.

ccf committed:
feat: Add Alpha and Beta metrics to backtesting

Add Alpha and Beta calculations to backtest summary metrics using CAPM model.

- Add calculate_alpha_beta() function in data/metrics.py
- Add helper functions _calculate_daily_returns() and _calculate_market_daily_returns()
- Include processed_data in backtest results for market benchmark calculation
- Add alpha and beta to strat_stats dictionary
- Display Alpha and Beta in SUMMARY METRICS table
- Fix typo: sharp_ratio -> sharpe_ratio

Alpha measures excess returns over market benchmark.
Beta measures strategy sensitivity to market movements.
Market benchmark is equal-weight buy-and-hold of pairlist.

ccf committed:
fix: Fix variable naming in hyperopt loss functions

- Fix typo: sharp_ratio -> sharpe_ratio for consistency
- Remove unnecessary blank line in sortino loss function

ccf committed:
test/docs: Add tests and documentation for Alpha/Beta metrics

Tests:
- Add test_calculate_alpha_beta() with real backtest data
- Add test_calculate_alpha_beta_synthetic() with parametrized cases
- Add test_calculate_alpha_beta_edge_cases() for empty/missing data
- Tests validate Beta correlation patterns and Alpha calculations

Documentation:
- Add detailed Alpha/Beta explanation to backtesting.md
- Explain Beta values (amplification, dampening, inverse)
- Explain Alpha as excess returns metric
- Provide example interpretation
- Note CAPM usage with risk-free rate of 0%

ccf committed:
docs: Improve Alpha/Beta documentation formatting

- Add periods to numbered list items for consistency
- Simplify note section into main text
- Improve text flow

ccf committed:
fix: Use portfolio value for alpha/beta returns calculation

The previous implementation calculated strategy returns by dividing
profits by the static starting balance, causing severe distortions
for strategies with different holding periods. As accounts grew or
shrank, returns appeared artificially inflated or deflated compared
to the market benchmark.

This fix properly tracks portfolio value over time and calculates
each day's return relative to the current portfolio value, ensuring
accurate compounding. This makes alpha/beta metrics meaningful and
comparable across all strategy types - from high-frequency scalping
to long-term position holding.

Changes:
- Modified _calculate_daily_returns() to track running portfolio value
- Added test_calculate_alpha_beta_compounding() to verify fix
- Updated documentation to explain the calculation method

ccf committed:
refactor: Use btdata instead of processed_data for alpha/beta calculation

Alpha/beta calculations only require OHLCV data (specifically the close
prices) to compute market returns. The processed_data was unnecessary
duplication as btdata already contains this information.

Changes:
- Remove processed_data from backtest results
- Pass btdata to generate_strategy_stats()
- Use btdata directly for alpha/beta calculation
- Reduces memory usage and storage size of backtest results
